File: Service/Service.cs
using Lab2C.Repository;
using Lab2C.Model;
using System.Collections.Generic;
using System;

namespace Lab2C.Service
{
    public class Service
    {
        private AdminRepository adminRepo;
        private BookingRepository bookingRepo;
        private TripRepository tripRepo;


        public Service(AdminRepository admin, BookingRepository booking, TripRepository trip)
        {
            this.adminRepo = admin;
            this.bookingRepo = booking;
            this.tripRepo = trip;
        }

        public bool Login(string username, string password)
        {
            Admin admin = adminRepo.AuthenticateAdmin(username, password);

            if (admin != null)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public IEnumerable<Trip> GetAllTrips()
        {
            return tripRepo.FindAll();
        }

        public IEnumerable<Booking> GetBookingsForTrip(long tripId)
        {
            return tripRepo.FindBookingsForTrip(tripId);
        }

        public void AddBooking(long tripId, string firstName, string lastName, int seats)
        {
            Trip trip = tripRepo.FindOne(tripId);

            Booking booking = new Booking(trip, seats, firstName, lastName);

            bookingRepo.Save(booking);
            trip.FreeSeats -= seats;

            tripRepo.Update(trip);
        }

        public IEnumerable<Trip> GetTripsBySourceAndDate(string source, DateTime date)
        {
            var res = tripRepo.FindTripsBySource(source);
            IList<Trip> result = new List<Trip>();

            //filer by data when data is fixed
            foreach(Trip tr in res)
            {
                if(tr.DepartureTime.Day == date.Day && tr.DepartureTime.Year == date.Year && tr.DepartureTime.Month == date.Month)
                {
                    result.Add(tr);
                }
            }


            return result;
        }
    }
}

File: Server/src/main/java/Service.java
/*
 *  @author albua
 *  created on 28/02/2021
 */

import java.security.NoSuchAlgorithmException;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * Service class for the application
 */
public class Service{
    private TripRepoInterface repoTrip;
    private AdminRepoInterface repoAdmins;
    private BookingRepoInterface repoBooking;

    public Service(TripRepoInterface repoTrip, AdminRepoInterface repoAdmins, BookingRepoInterface
            repoBooking) {
        this.repoTrip = repoTrip;
        this.repoAdmins = repoAdmins;
        this.repoBooking = repoBooking;
    }

    public Iterable<Trip> getAllTrips(){

        return repoTrip.findAll();
    }

    public Iterable<Booking> getBookingsForTrip(Long tripId){

        return repoTrip.findBookingsForTrip(tripId);
    }

    public void addBooking(Long tripId, String firstName, String lastName, int seats){
        Trip trip = repoTrip.findOne(tripId);

        Booking booking = new Booking(trip, seats, firstName, lastName);

        repoBooking.save(booking);
        trip.setFreeSeats(trip.getFreeSeats() - seats);
        repoTrip.update(trip);
    }

    public Iterable<Trip> getTripsBySourceAndDate(String source, LocalDateTime date){
        Iterable<Trip> trips = repoTrip.findTripsBySource(source);

        HashSet<Trip> result = new HashSet<>();
        trips.forEach(x->{
            if(date.getDayOfMonth() == x.getDepartureTime().getDayOfMonth() && date.getMonth() == x.getDepartureTime().getMonth()
            && date.getYear() == x.getDepartureTime().getYear()){
                result.add(x);
            }
        });

        return result;
    }

    public boolean loginHandler(String username, String password){
        try {
            return repoAdmins.authenticateAdmin(username, password) != null;
        }
        catch (NoSuchAlgorithmException e){
            e.printStackTrace();
        }
        return false;
    }

    public Admin findOneAdminUsername(String username){
        Iterable<Admin> admins = repoAdmins.findAll();

        List<Admin> lAdmin = StreamSupport.stream(admins.spliterator(), false).collect(Collectors.toList());

        for (Admin admin : lAdmin) {
            if (admin.getUsername().equals(username)) {
                return admin;
            }
        }

        return null;
    }
}
